File: P/841_Keys_and_Rooms.java
class Solution {
    List<List<Integer>> rooms;
    Stack<Integer> keys;
    boolean[] visited;
    public boolean canVisitAllRooms(List<List<Integer>> rooms) {
        this.rooms = rooms;
        keys = new Stack();
        visited = new boolean[rooms.size()];

        visit(0);
        while(!keys.isEmpty()) 
            visit(keys.pop());
        
        for(int i = 0; i < visited.length; i++) {
            if(!visited[i])
                return false;
        }
        
        return true;
    }
    
    private void visit(int room) {
        visited[room] = true;
        for(Object n: this.rooms.get(room).toArray()) {
            int r = (int) n;
            if(!visited[r])
                keys.push(r);
        }
    }
}

File: 주영/Graph/841. Keys and Rooms.js
var canVisitAllRooms = function (rooms) {
  const uniqueKeys = new Set([0]); // 어차피 중복된 키는 소용이 없기 때문에 유일한 값을 저장하기 위해 set을 사용한다.
  const stack = [0]; // 0번째 키는 반드시 있기 때문에 0을 넣어준다.

  while (stack.length) {
    let currentKey = stack.pop(); // stack의 가장 위에 있는 key를 pop해주고

    rooms[currentKey].forEach(key => {
      if (!uniqueKeys.has(key)) {
        // 해당 키가 set에 없던 키라면
        uniqueKeys.add(key); // uniqueKeys에도 넣어주고
        stack.push(key); // stack에도 넣어준다.
      }
    });
  }

  // 결국 uniqueKeys랑 rooms 배열의 길이가 같으면 그게 답이 된다.
  return uniqueKeys.size === rooms.length;
};

let rooms = [[1, 3], [3, 0, 1], [2], [0]];
canVisitAllRooms(rooms);
